### PR TITLE
fix: [DHIS2-9023] Block programStage deletion when a program rule is linked to it

### PR DESCRIPTION
**Pull request by enricocolasante**

### CODE REVIEW

**Review by stian-sandvold — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleDeletionHandler.java [hunk 1]
```diff
@@ -72,4 +78,31 @@ public void deleteProgram( Program program )
...
+
+        if ( programRuleVariables.isEmpty() )
+        {
+            return null;
+        }
+
+        return programRuleVariables.stream()
+            .map( BaseIdentifiableObject::getName )
+            .collect( Collectors.joining( ", " ) );
+    }
```
Comment (gnespolino) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleDeletionHandler.java:99]: Minor: Might be slightly more compact and avoid to stream twice, like this:
 
```suggestion
    @Override
    public String allowDeleteProgramStage( ProgramStage programStage )
    {
        String programRuleVariables = programRuleService
            .getProgramRule( programStage.getProgram() )
            .stream()
            .filter( pr -> isLinkedToProgramStage( programStage, pr ) )
            .map( BaseIdentifiableObject::getName )
            .collect( Collectors.joining( ", " ) );

        return StringUtils.isBlank( programRuleVariables ) ? null : programRuleVariables;
    }
```

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleVariableDeletionHandler.java [hunk 2]
```diff
@@ -61,6 +67,25 @@ protected String getClassName()
...
+
+        if ( programRuleVariables.isEmpty() )
+        {
+            return null;
+        }
+
+        return programRuleVariables.stream()
+            .map( BaseIdentifiableObject::getName )
+            .collect( Collectors.joining( ", " ) );
+    }
```
Comment (gnespolino) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleVariableDeletionHandler.java:88]: Same comment as before.

**Review by gnespolino — APPROVED** [review]
You might apply a minor change if you feel it's better.

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleVariableDeletionHandler.java [hunk 3]
```diff
@@ -61,6 +67,25 @@ protected String getClassName()
     {
         return ProgramRuleVariable.class.getSimpleName();
     }
+
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleVariableDeletionHandler.java:70]:
```suggestion
@Override
    public String allowDeleteProgramStage( ProgramStage programStage )
    {
        List<ProgramRuleVariable> programRuleVariables = programRuleVariableService
            .getProgramRuleVariable( programStage.getProgram() )
            .stream()
            .filter( prv -> Objects.equals( prv.getProgramStage(), programStage ) )
            .map( BaseIdentifiableObject::getName )
            .collect( Collectors.joining( ", " ) );

        return StringUtils.isBlank( programRuleVariables ) ? null : programRuleVariables;
    }
```

Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleVariableDeletionHandler.java:70]:
```suggestion
@Override
    public String allowDeleteProgramStage( ProgramStage programStage )
    {
        List<ProgramRuleVariable> programRuleVariables = programRuleVariableService
            .getProgramRuleVariable( programStage.getProgram() )
            .stream()
            .filter( prv -> Objects.equals( prv.getProgramStage(), programStage ) )
            .map( BaseIdentifiableObject::getName )
            .collect( Collectors.joining( ", " ) );

        return StringUtils.isBlank( programRuleVariables ) ? null : programRuleVariables;
    }
```

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleVariableDeletionHandler.java (hunk 2 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleVariableDeletionHandler.java]: Done

**Pull request merged**